### ISSUE TITLE
`set throttle` IPS value second-guessing

### ISSUE THREAD

**Issue by tangentsoft** [simh/simh#429]
Hello, PiDP-8/I maintainer here again!

Before I continue, I want to say that this is just an FYI bug report, because I've already worked out a solution that works for our purposes, which is to run the simulator as fast as possible but leave about half the host CPU power left for background tasks. (`set throttle 50%`) I just thought you'd like to think about these things, if you weren't already aware of them.

My prior effort at achieving that end was to try and find the ideal IPS value for the host platform. On a Raspberry Pi B+, I was able to get up to about `set throttle 850k` before SIMH started complaining about insufficient host CPU power. This was when starting with the `set throttle` command in the simulator startup configuration file, and when starting from a host idle state.

I found that if I waited until after the simulator is up to start giving `set throttle` commands, I could get it up to about 1.1 MIPS on that same host before it starts complaining of insufficient host CPU power. Why the difference? Why do I lose ~0.3 MIPS just because I set the throttle from the start instead of waiting a few seconds?

I then found that if I started the simulator without throttling, I got about 2.4 MIPS on that same host according to `show clock`. Why is this 2x over the limit it claims to have hit with manual throttling?

With my `set throttle 50%` value, it's indeed about half that, 1.2 MIPS, so it is not the case that the very mechanism of throttling soaks up the difference.

This problem came to a head when I discovered that my efforts at pushing the throttle upward toward some hypothetical limit were for naught when it came to starting the simulator on Pi boot, since systemd runs multiple processes in parallel, thus reducing host CPU power available to SIMH. That meant I needed to lower IPS even further to avoid causing SIMH to run unthrottled, hogging the whole CPU, just because it was a bit resource constrained at simulator startup time.

Maybe you do nothing about any of this, since as I said, there is a solution in the current code.

Before you dismiss this, please reconsider whether `set nothrottle` is a good response to "not enough CPU power." That just exacerbates the diagnosed problem, irrespective of whether the diagnosis was correct or not.

How about this for an alternate plan:

1. Don't judge whether the host CPU is oversubscribed in the first second of simulator run time. Wait some seconds to see if things stabilize.

2. If it is still oversubscribed by that time, drop the requested IPS value down by some amount. Say, -10%. Check back in a second later, and keep ratcheting it downward once per second until you can reliably achieve the new rate.

3. Complain only after several ratchet steps downward, but keep trying.

**Comment by markpizz** [comment]
I suspect that the latest throttling changes will be closer to what you've asked for.  Let me know.

**Comment by markpizz** [comment]
If there is still system activity at boot time which competes with simh's throttle measurements, you can add a sleep command to your simh startup script before you boot the simulated system:

      sleep 45s
      set throttle 1200K
      boot dev

The current throttle logic runs all out for THROT_DELAY seconds (default 20) before it actually determines the initial throttling parameters.

You could:

      set throt 333k
      deposit int-throttle THROT_DELAY 10
      boot dev

**Comment by tangentsoft** [comment]
> I suspect that the latest throttling changes will be closer to what you've asked for. Let me know.

To test this, I tried `set throttle 50M` on a Pi 3, which is impossible. The actual max is around 8 MIPS with the wind at our backs.

I did indeed not immediately get a notification from SIMH that it was going to ignore my chosen throttle value, but then 20 seconds later, the new throttling behavior kicked in and it told me

    Host CPU can only simulate 7,610,400 instructions per second

So, it does seem able to figure things out better now. 

Between that and the ability to hold off the decision about whether the CPU is oversubscribed, I think this does solve the problem. Thanks!

I haven't tested this thoroughly yet, but because it looks like the right thing, I'm going to close this issue now. If I'm surprised by my later testing, I can reopen this issue or file a new one, depending on what I find.

> you can add a sleep command to your simh startup script

Noted. I think we'll just set `THROT_DELAY` go to a higher value in this case, overriding it in the interactive case only, when the system should be idle at simulator startup.
